File: memcontainers/hosts/js/src/host_call.ts
// Host-call capability: the TS host side of `mc_sys_host_call`. The kernel hands the host an opaque
// request blob (`name\0args`, from `mc-tool`); the host routes it to a registered handler and streams
// back a result. Poll-based, mirroring the net capability — and async-friendly, since tool handlers
// are async: `start` kicks off the handler, `poll` reports `0` until it resolves, then `body` streams
// the result. Default-deny (A9).

export interface HostCallCapability {
  /** Start a call; return an opaque handle >= 0, or -1 to refuse. */
  start(req: Uint8Array): number;
  /** `0` in flight, `1` ready, `-1` failed/unknown. */
  poll(handle: number): number;
  /** Result bytes into `buf`: `n > 0`, `0` = EOF, `-1` = error. */
  body(handle: number, buf: Uint8Array): number;
  close(handle: number): void;
}

/** Default: refuse every host call (no tools installed). */
export class DeniedHostCall implements HostCallCapability {
  start(): number {
    return -1;
  }
  poll(): number {
    return -1;
  }
  body(): number {
    return -1;
  }
  close(): void {}
}

/** A handler for a named tool: given the args string (typically JSON), return a result (text or
 *  bytes). May be async. */
export type ToolHandler = (args: string) => Promise<Uint8Array | string> | Uint8Array | string;

/** A binary-safe handler: given the request body (the bytes after `name\0`, verbatim — no UTF-8
 *  decode, no trailing-NUL trim), return a result. Used by host-backed mount drivers, whose WRITE op
 *  carries binary file content. */
export type RawToolHandler = (body: Uint8Array) => Promise<Uint8Array | string> | Uint8Array | string;

interface Slot {
  result?: Uint8Array;
  offset: number;
  failed: boolean;
  done: boolean;
}

/** A registry of named (possibly async) handlers — the host-call capability for the JS host, driven by
 *  `vm.tool` (UTF-8 `tools`) and `vm.mount` (binary-safe `raw` handlers, keyed by absolute mount
 *  path). */
export class MapHostCall implements HostCallCapability {
  private readonly tools = new Map<string, ToolHandler>();
  private readonly raw = new Map<string, RawToolHandler>();
  private readonly slots = new Map<number, Slot>();
  private next = 1;

  register(name: string, handler: ToolHandler): void {
    this.tools.set(name, handler);
  }
  /** Register a binary-safe handler (a host-backed mount, keyed by its absolute path). Checked before
   *  the UTF-8 `tools` map. */
  registerRaw(name: string, handler: RawToolHandler): void {
    this.raw.set(name, handler);
  }
  unregister(name: string): void {
    this.tools.delete(name);
    this.raw.delete(name);
  }
  has(name: string): boolean {
    return this.tools.has(name) || this.raw.has(name);
  }

  /** Allocate a slot and run `produce` (sync or async); the result streams back via `body` once it
   *  resolves. */
  private startSlot(produce: () => Promise<Uint8Array | string> | Uint8Array | string): number {
    const handle = this.next;
    this.next = this.next + 1 < 1 ? 1 : this.next + 1;
    const slot: Slot = { offset: 0, failed: false, done: false };
    this.slots.set(handle, slot);
    Promise.resolve()
      .then(produce)
      .then((res) => {
        slot.result = typeof res === "string" ? new TextEncoder().encode(res) : res;
        slot.done = true;
      })
      .catch(() => {
        slot.failed = true;
        slot.done = true;
      });
    return handle;
  }

  start(req: Uint8Array): number {
    // Mount drivers register binary-safe handlers keyed by absolute path; tools register UTF-8
    // handlers keyed by bare name. The two key spaces are disjoint (mount names start with `/`), so
    // check raw first — passing the body verbatim (no UTF-8 decode, no NUL-trim, since a binary WRITE
    // may end in 0x00).
    const rawNul = req.indexOf(0);
    const rawName = new TextDecoder().decode(rawNul < 0 ? req : req.subarray(0, rawNul));
    const rawHandler = this.raw.get(rawName);
    if (rawHandler) {
      const body = rawNul < 0 ? new Uint8Array(0) : req.subarray(rawNul + 1);
      return this.startSlot(() => rawHandler(body));
    }

    // argv blobs are NUL-terminated; trim trailing NUL(s) before splitting so the args string doesn't
    // carry a stray `\0` that would break JSON parsing.
    let trimEnd = req.length;
    while (trimEnd > 0 && req[trimEnd - 1] === 0) trimEnd--;
    req = req.subarray(0, trimEnd);
    const nul = req.indexOf(0);
    const name = new TextDecoder().decode(nul < 0 ? req : req.subarray(0, nul));
    const args = nul < 0 ? "" : new TextDecoder().decode(req.subarray(nul + 1));
    const handler = this.tools.get(name);
    if (!handler) return -1;
    return this.startSlot(() => handler(args));
  }

  poll(handle: number): number {
    const s = this.slots.get(handle);
    if (!s) return -1;
    if (s.failed) return -1;
    return s.done ? 1 : 0;
  }

  body(handle: number, buf: Uint8Array): number {
    const s = this.slots.get(handle);
    if (!s || s.failed || !s.result) return -1;
    const remaining = s.result.subarray(s.offset);
    const n = Math.min(remaining.length, buf.length);
    buf.set(remaining.subarray(0, n), 0);
    s.offset += n;
    return n;
  }

  close(handle: number): void {
    this.slots.delete(handle);
  }
}

File: memcontainers/hosts/wasmtime/src/host_call.rs
//! Host-call capability: the host side of `mc_sys_host_call`. The kernel hands
//! the host an opaque request blob (`name\0args`, from the tool broker); the host
//! routes it to a registered handler and streams back a result. Poll-based,
//! mirroring the net capability: `start` → handle, `poll` → readiness, `body` →
//! result bytes, `close`. Default-deny (A9): without an installed capability,
//! every call is refused, so host tools are unavailable to a VM unless the host
//! opts in.

use std::collections::HashMap;

/// The host-side host-call surface, gated like `NetCapability`.
pub trait HostCallCapability: Send + 'static {
    /// Start a call from the request blob; return an opaque handle ≥ 0, or -1 to
    /// refuse (no handler / denied).
    fn start(&mut self, req: &[u8]) -> i32;
    /// Readiness: `0` in flight, `1` ready, `-1` failed/unknown.
    fn poll(&mut self, handle: i32) -> i32;
    /// Stream result bytes into `buf`: `n > 0` bytes, `0` = EOF, `-1` = error.
    fn body(&mut self, handle: i32, buf: &mut [u8]) -> i32;
    /// Release the handle.
    fn close(&mut self, handle: i32);
}

/// Default: refuse every host call (no tools installed).
pub struct DeniedHostCall;

impl HostCallCapability for DeniedHostCall {
    fn start(&mut self, _req: &[u8]) -> i32 {
        -1
    }
    fn poll(&mut self, _handle: i32) -> i32 {
        -1
    }
    fn body(&mut self, _handle: i32, _buf: &mut [u8]) -> i32 {
        -1
    }
    fn close(&mut self, _handle: i32) {}
}

/// A synchronous handler: given the args string, return a result or an error.
pub type ToolFn = Box<dyn FnMut(&str) -> Result<Vec<u8>, String> + Send>;

/// A binary-safe handler: given the request body (the bytes after `name\0`,
/// verbatim — no UTF-8 decode, no trailing-NUL trim), return a result. Used by
/// host-backed mount drivers, whose WRITE op carries binary file content.
pub type RawToolFn = Box<dyn FnMut(&[u8]) -> Result<Vec<u8>, String> + Send>;

struct Slot {
    result: Vec<u8>,
    offset: usize,
    failed: bool,
}

/// A registry of named synchronous handlers — the host-call capability for the
/// native host (tests, the in-process embedded path). Handlers run immediately
/// in `start`, so `poll` reports ready at once.
pub struct MapHostCall {
    tools: HashMap<String, ToolFn>,
    raw: HashMap<String, RawToolFn>,
    slots: HashMap<i32, Slot>,
    next: i32,
}

impl MapHostCall {
    pub fn new() -> Self {
        Self {
            tools: HashMap::new(),
            raw: HashMap::new(),
            slots: HashMap::new(),
            next: 1,
        }
    }

    /// Register a handler for `name` (the first NUL-separated field of the blob).
    pub fn register(&mut self, name: &str, f: ToolFn) {
        self.tools.insert(name.to_string(), f);
    }

    /// Register a binary-safe handler for `name` (a host-backed mount, keyed by
    /// its absolute mount path). Checked before the UTF-8 `tools` map.
    pub fn register_raw(&mut self, name: &str, f: RawToolFn) {
        self.raw.insert(name.to_string(), f);
    }

    /// Drop a previously-registered handler (mount unmounted / tool removed).
    pub fn unregister(&mut self, name: &str) {
        self.tools.remove(name);
        self.raw.remove(name);
    }
}

impl Default for MapHostCall {
    fn default() -> Self {
        Self::new()
    }
}

/// Split a blob into `(name, body)` without trimming trailing NULs — the name is
/// the bytes up to the first NUL, the body is everything after it, verbatim. Used
/// for binary-safe (mount) handlers, where the WRITE body can legitimately end in
/// `0x00`.
fn split_req_raw(req: &[u8]) -> (&str, &[u8]) {
    let nul = req.iter().position(|&b| b == 0).unwrap_or(req.len());
    let name = std::str::from_utf8(&req[..nul]).unwrap_or("");
    let body = if nul < req.len() {
        &req[nul + 1..]
    } else {
        &[]
    };
    (name, body)
}

fn split_req(req: &[u8]) -> (&str, &str) {
    // argv blobs are NUL-terminated, so trim any trailing NUL(s) before splitting
    // — otherwise the args string carries a stray `\0` that breaks JSON parsing.
    let end = req
        .iter()
        .rposition(|&b| b != 0)
        .map(|i| i + 1)
        .unwrap_or(0);
    let req = &req[..end];
    let nul = req.iter().position(|&b| b == 0).unwrap_or(req.len());
    let name = std::str::from_utf8(&req[..nul]).unwrap_or("");
    let args = if nul < req.len() {
        std::str::from_utf8(&req[nul + 1..]).unwrap_or("")
    } else {
        ""
    };
    (name, args)
}

impl HostCallCapability for MapHostCall {
    fn start(&mut self, req: &[u8]) -> i32 {
        // Mount drivers register binary-safe handlers keyed by absolute path; tools
        // register UTF-8 handlers keyed by bare name. The two key spaces are
        // disjoint (mount names start with `/`), so checking raw first is safe.
        let (raw_name, raw_body) = split_req_raw(req);
        if self.raw.contains_key(raw_name) {
            let body = raw_body.to_vec();
            let f = self.raw.get_mut(raw_name).expect("just checked");
            let handle = self.next;
            self.next = self.next.wrapping_add(1).max(1);
            let slot = match f(&body) {
                Ok(result) => Slot {
                    result,
                    offset: 0,
                    failed: false,
                },
                Err(_) => Slot {
                    result: Vec::new(),
                    offset: 0,
                    failed: true,
                },
            };
            self.slots.insert(handle, slot);
            return handle;
        }

        let (name, args) = split_req(req);
        let Some(f) = self.tools.get_mut(name) else {
            return -1; // unknown tool → refuse
        };
        let handle = self.next;
        self.next = self.next.wrapping_add(1).max(1);
        let slot = match f(args) {
            Ok(result) => Slot {
                result,
                offset: 0,
                failed: false,
            },
            Err(_) => Slot {
                result: Vec::new(),
                offset: 0,
                failed: true,
            },
        };
        self.slots.insert(handle, slot);
        handle
    }

    fn poll(&mut self, handle: i32) -> i32 {
        match self.slots.get(&handle) {
            Some(s) if s.failed => -1,
            Some(_) => 1,
            None => -1,
        }
    }

    fn body(&mut self, handle: i32, buf: &mut [u8]) -> i32 {
        match self.slots.get_mut(&handle) {
            Some(s) if s.failed => -1,
            Some(s) => {
                let remaining = &s.result[s.offset..];
                let n = remaining.len().min(buf.len());
                buf[..n].copy_from_slice(&remaining[..n]);
                s.offset += n;
                n as i32
            }
            None => -1,
        }
    }

    fn close(&mut self, handle: i32) {
        self.slots.remove(&handle);
    }
}
